hkm/views/views.py

```diff
@@ -475,27 +475,61 @@ def get_template_names(self):
+
+        url_params = request.session.get('url_params', {})
+        search_term_changed = self.url_params['search'] != url_params.get('search', '')
+        page_changed = self.url_params['page'] != url_params.get('page')
+        author_changed = self.url_params['author'] != url_params.get('author')
+        date_changed = self.url_params['date'] != url_params.get('date')
+        date_from_changed = self.url_params['date_from'] != url_params.get('date_from')
+        date_to_changed = self.url_params['date_to'] != url_params.get('date_to')
+
+        session_facet_result = request.session.get('facet_result', None)
+
+        if not session_facet_result:
+            self.facet_result = self.get_facet_result(self.url_params['search'])
+            # date_from & date_to require full list of dates
+            self.all_dates = self.facet_result.get('facets', {}).get('main_date_str', [])
+        # If search term or year to year changed => fetch facets again
+        elif search_term_changed or date_from_changed or date_to_changed:
+            self.facet_result = self.get_facet_result(self.url_params['search'], self.url_params['date_from'], self.url_params['date_to'])
+            self.all_dates = request.session.get('all_dates')
+        else:
+            self.facet_result = request.session.get('facet_result')
+            self.all_dates = request.session.get('all_dates')
+
+        facets = {}
+        if self.url_params['author']:
+            facets['author_facet'] = self.url_params['author']
+        if self.url_params['date']:
+            facets['main_date_str'] = self.url_params['date']
+        if self.url_params['date_from'] or self.url_params['date_to']:
+            facets['search_daterange_mv'] = "[%s TO %s]" % (self.url_params['date_from'], self.url_params['date_to'])
+
         load_all_pages = bool(int(request.GET.get('loadallpages', 1)))
 
         session_search_result = copy.deepcopy(request.session.get('search_result', {}))
-        records = session_search_result.get('records', []) if self.search_term != request.GET.get('search', '') else []
+        records = session_search_result.get('records', []) if search_term_changed else []
 
         # If we don't hit any checks below this point search_results end up being empty
         self.search_result = session_search_result
 
-        search_term_changed = self.search_term != request.session.get('search')
-        page_changed = self.page != request.session.get('page')
-
         # This if statement is true when user makes search with new term or parameters in list view
         if load_all_pages and not kwargs.get('record'):
-            if search_term_changed or page_changed:
-                results = self.get_search_result(self.search_term, self.page, self.page_size)
+            if search_term_changed or page_changed or author_changed or date_changed or date_from_changed or date_to_changed:
```

Comment:
Instead of making the big function even bigger, there could be a couple of private functions like __has_search_params_changesand __update_search_params that takes care of comparing the current situation to the previous one and updating the params.

Reply:
Yup, this is something I'm taking a look right now. It seems that I might have missed at least one more place where similar check needs to be done and writing this kind of check every time makes complicates things.